restructure managed jobs docs (#4681)
skypilot-org/skypilot: docs/source/examples/managed-jobs.rst

Reviewer: For https://docs.skypilot.co/en/managed-jobs-docs-revamp/examples/managed-jobs.html#intermediate-storage-for-files
- Sentence Case to be consistent?
- Some ideas to simplify the section name: Temporary Bucket; Setting the Job Files Bucket.  
  - Sort of feeling 'intermediate bucket' can be hard to parse

----

Useful CLIs and Job Dashboard feel like two short utility sections that should sit next to each other. Maybe pull them up before managed pipelines.

   processing, training a model, and then running inference on it.

Create a Managed Job
--------------------

A managed job is created from a standard :ref:`SkyPilot YAML <yaml-spec>`.

.. tip::
   You can test your YAML on unamanged :code:`sky launch`, then do a production run as a managed job using :code:`sky jobs launch`.

> For example, we found the BERT fine-tuning YAML works with :code:`sky launch`, and want to

Reviewer: Rephrase a bit to be more natural reading?

.. note::

  :ref:`workdir <sync-code-artifacts>` and :ref:`file mounts with local files <sync-code-artifacts>` will be :ref:`automatically uploaded to a cloud bucket <intermediate-bucket>`.
  The bucket will be cleaned up after the job finishes.

To launch this YAML as a managed job, use :code:`sky jobs launch`:

.. code-block:: console

  $ sky jobs launch -n bert-qa-job bert_qa.yaml

:code:`sky jobs launch` works similarly to :code:`sky launch` - you can run :code:`sky jobs launch --help` or see the :ref:`CLI reference <sky-job-launch>` for more information.

SkyPilot will launch and start monitoring the job.

- Under the hood, SkyPilot spins up a temporary cluster for the job.

> - If a spot preemption or any machine failure happens, SkyPilot will automatically search for resources across regions and clouds to re-launch the job.

Author: Strong disagree. The "search space" of a job is not a concept that will be naturally obvious to people. We could add something like "search ... across regions and clouds _(within the specified constraints)_ to..." but I think it confuses the sentence too much.

- Resources are cleaned up as soon as the job is finished.

.. tip::
   You can test your YAML on |unmanaged sky launch|_ , then do a production run as a managed job using :code:`sky jobs launch`.

.. https://stackoverflow.com/a/4836544
.. |unmanaged sky launch| replace:: unmanaged :code:`sky launch`
.. _unmanaged sky launch: ../getting-started/quickstart.html

:code:`sky launch` and :code:`sky jobs launch` have a similar interface, but are useful in different scenarios.

.. list-table::
   :header-rows: 1

   * - :code:`sky launch` (cluster jobs)
     - :code:`sky jobs launch` (managed jobs)
   * - Long-lived, manually managed cluster
     - Dedicated auto-managed cluster for each job
   * - Spot preemptions must be manually recovered
     - Spot preemptions are auto-recovered

>    * - Useful for launching and working with a single cluster/job

Author: Will pass on this change. The goal is just to describe the recommended workflow, not the actual technical constraints.

>    * - Useful for launching and working with a single cluster/job

Reviewer: Discussed offline; we should find a way.

>    * - Useful for launching and working with a single cluster/job

Author: @concretevitamin how is
```
   * - Job parallelism limited by cluster resources
     - Easily manage hundreds or thousands of jobs at once
```

>    * - Useful for launching and working with a single cluster/job

Reviewer: Job parallelism may refer to number of threads/tasks within a job.

>    * - Useful for launching and working with a single cluster/job

Author: Just "parallelism" then? I don't really like "Number of parallelism"

>    * - Useful for launching and working with a single cluster/job

Author: @concretevitamin, updated, ready for review

See a list of all managed jobs:

.. code-block:: console

  $ sky jobs queue

.. code-block:: console

  Fetching managed jobs...
  Managed jobs:
  ID NAME     RESOURCES           SUBMITTED   TOT. DURATION   JOB DURATION   #RECOVERIES  STATUS
  2  roberta  1x [A100:8][Spot]   2 hrs ago   2h 47m 18s      2h 36m 18s     0            RUNNING
  1  bert-qa  1x [V100:1][Spot]   4 hrs ago   4h 24m 26s      4h 17m 54s     0            RUNNING

Stream the logs of a running managed job:

.. code-block:: console

  $ sky jobs logs -n bert-qa  # by name
  $ sky jobs logs 2           # by job ID

Cancel a managed job:

.. code-block:: console

  $ sky jobs cancel -n bert-qa  # by name
  $ sky jobs cancel 2           # by job ID

.. note::
  If any failure happens for a managed job, you can check :code:`sky jobs queue -a` for the brief reason
  of the failure. For more details, it would be helpful to check :code:`sky jobs logs --controller <job_id>`.

Jobs dashboard
~~~~~~~~~~~~~~

Use ``sky jobs dashboard`` to open a dashboard to see all jobs:

.. code-block:: console

  $ sky jobs dashboard

This automatically opens a browser tab to show the dashboard:

> .. image:: ../images/job-dashboard.png

Author: yes, it was on quickstart page

   In exchange, spot instances are significantly cheaper than normal instances that are not subject to preemption (so-called "on-demand" instances).
   Depending on the cloud and VM type, spot instances can be 70-90% cheaper.

SkyPilot automatically finds available spot instances across regions and clouds to maximize availability.
Any spot preemptions are automatically handled by SkyPilot without user intervention.

.. note::
   By default, a job will be restarted from scratch after each preemption recovery.
   To avoid redoing work after recovery, implement :ref:`checkpointing and recovery <checkpointing>`.
   Your application code can checkpoint its progress periodically to a :ref:`mounted cloud bucket <sky-storage>`. The program can then reload the latest checkpoint when restarted.

Here is an example of a training job failing over different regions across AWS and GCP.

.. image:: https://i.imgur.com/Vteg3fK.gif
  :width: 600
  :alt: GIF for BERT training on Spot V100

.. image:: ../images/spot-training.png
  :width: 600
  :alt: Static plot, BERT training on Spot V100

Quick comparison between *managed spot jobs* vs. *launching unmanaged spot clusters*:

> .. list-table::

Reviewer: For discussion: Does this feel overlapping with the new table above too much?

   :widths: 30 18 12 35
   :header-rows: 1

   * - Command
     - Managed?
     - SSH-able?
     - Best for
   * - :code:`sky jobs launch --use-spot`
     - Yes, preemptions are auto-recovered
     - No
     - Scaling out long-running jobs (e.g., data processing, training, batch inference)
   * - :code:`sky launch --use-spot`
     - No, preemptions are not handled
     - Yes

>      - Interactive dev on spot instances (especially for hardware with low preemption rates)

Author: a bit, @concretevitamin said the same https://github.com/skypilot-org/skypilot/pull/4681/files#r1951770844

I think it's not terrible. It could be better but I will not spend time on it right now.

  $ sky jobs launch -n bert-qa-job bert_qa.yaml

> :code:`sky jobs launch` has a similar interface to :code:`sky launch` - you can run :code:`sky jobs launch --help` or see the :ref:`CLI reference <sky-job-launch>` for more information.

Reviewer: Still not sure about keeping this para. Some major arguments are missing in `jobs launch`, so this statement isn't 100% true. Still repeats a para below.

Author: ok will just delete it

> SkyPilot will launch and start monitoring the job. When a spot preemption or any machine failure happens, SkyPilot will automatically

Reviewer: L112 should refer to the new https://docs.skypilot.co/en/managed-jobs-docs-revamp/examples/managed-jobs.html#intermediate-storage-for-files support briefly?

search for resources across regions and clouds to re-launch the job.

.. _spot-jobs:

> Running on Spot

Reviewer: ```suggestion
Running on Spot Instances
```

---------------

Managed jobs can run on spot instances, and preemptions are auto-recovered by SkyPilot.

.. tip::
   Spot instances are cloud VMs that may be "preempted".
   The cloud provider can forcibly shut down the underlying VM and remove your access to it, interrupting the job running on that instance.

>    This can be inconvenient, but SkyPilot will automatically restart/recover your job on another instance.

Reviewer: ```suggestion
   When preemptions happen, a SkyPilot managed job will automatically recover and restart on another temporary cluster.
```
Or, seems ok to remove, 2nd para last sentence says similar.

>    In exchange for unpredictability, spot instances are significantly cheaper than normal instances that are not subject to preemption (so-called "on-demand" instances).

Reviewer: ```suggestion
   In exchange for preemption, spot instances are significantly cheaper than normal instances that are not subject to preemption (so-called "on-demand" instances).
```

   Depending on the cloud and VM type, spot instances can be 70-90% cheaper.
   Since SkyPilot automatically handles preemptions, you can get these savings for free.

> To run on spot instances, use :code:`sky jobs launch --use-spot`, or specify :code:`use_spot: true` in your SkyPilot YAML.

Reviewer: Pull this up before the tip above?

Reviewer: Also, this para feels long; is there a way you see to break it up / simplify a little?

SkyPilot automatically finds available spot instances across regions and clouds to maximize availability.
Any spot preemptions are automatically handled by SkyPilot without user intervention.

.. tip::

>    The job will be restarted from scratch after each preemption recovery.

Reviewer: ```suggestion
   By default, a job will be restarted from scratch after each preemption recovery.
```

This seems important enough to be pull out from a tip box, not feeling strongly.

Author: My impression is that people are more likely to read tip/note boxes than things outside of the box, which is why I think this is useful. Maybe `note` is better.

   To avoid redoing work after recovery, implement :ref:`checkpointing and recovery <checkpointing>`.
   Your application code can checkpoint its progress periodically to a :ref:`mounted cloud bucket <sky-storage>`. The program can then reload the latest checkpoint when restarted.

> Here is an example of a BERT training job failing over different regions across AWS and GCP.

Reviewer: ```suggestion
Here is an example of a training job failing over different regions across AWS and GCP.
```

      mode: MOUNT

The :code:`MOUNT` mode in :ref:`SkyPilot bucket mounting <sky-storage>` ensures the checkpoints outputted to :code:`/checkpoint` are automatically synced to a persistent bucket.

Note that the application code should save program checkpoints periodically and reload those states when the job is restarted.
This is typically achieved by reloading the latest checkpoint at the beginning of your program.

.. _failure-recovery:

Jobs Restarts on User Code Failure
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Preemptions or hardware failures will be auto-recovered, but **by default, user code failures (non-zero exit codes) are not auto-recovered**.

In some cases, you may want a job to automatically restart even if it fails in application code. For instance, if a training job crashes due to an NVIDIA driver issue or NCCL timeout, it should be recovered. To specify this, you
can set :code:`max_restarts_on_errors` in :code:`resources.job_recovery` in the job YAML file.

.. code-block:: yaml

  resources:
    accelerators: A100:8
    job_recovery:
      # Restart the job up to 3 times on user code errors.
      max_restarts_on_errors: 3

> More advanced policies for resource selection, such as the `Can't Be Late
> <https://www.usenix.org/conference/nsdi24/presentation/wu-zhanghao>`__ (NSDI'24)
> paper, may be supported in the future.
> 
> 

Reviewer: ```suggestion
```

      mode: MOUNT

The :code:`MOUNT` mode in :ref:`SkyPilot bucket mounting <sky-storage>` ensures the checkpoints outputted to :code:`/checkpoint` are automatically synced to a persistent bucket.

To implement checkpointing in your application code:

1. Periodically save checkpoints containing the current program state during execution.
2. When starting/restarting, check if any checkpoints are present, and load the latest checkpoint.

> Checkpointing features are included in most model training libraries, such as `PyTorch <https://pytorch.org/docs/stable/checkpoint.html>` and `TensorFlow <https://www.tensorflow.org/guide/checkpoint>`.

Author: Is this text reasonable? I don't really know what I'm talking about when it comes to ML libraries.

Reviewer: This is correct :-) Suggested.

> Checkpointing features are included in most model training libraries, such as `PyTorch <https://pytorch.org/docs/stable/checkpoint.html>` and `TensorFlow <https://www.tensorflow.org/guide/checkpoint>`.

Reviewer: ```suggestion
Both features are included in most model training libraries. See below [TODO: link to the Bert example] for an example.
```

For other types of workloads, you can implement a similar mechanism as long as you can store the program state to/from disk.

.. _failure-recovery:

Jobs restarts on user code failure
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Preemptions or hardware failures will be auto-recovered, but **by default, user code failures (non-zero exit codes) are not auto-recovered**.

In some cases, you may want a job to automatically restart even if it fails in application code. For instance, if a training job crashes due to an NVIDIA driver issue or NCCL timeout, it should be recovered. To specify this, you
can set :code:`max_restarts_on_errors` in :code:`resources.job_recovery` in the job YAML file.

.. code-block:: yaml

  resources:
    accelerators: A100:8
    job_recovery:
      # Restart the job up to 3 times on user code errors.
      max_restarts_on_errors: 3

This will restart the job, up to 3 times (for a total of 4 attempts), if your code has any non-zero exit code. Each restart runs on a newly provisioned temporary cluster.

> Here's how various kinds of failures will be handled by SkyPilot:

Reviewer: This seems to warrant a subsection, "Recovery" / "Managed job recovery"?

.. list-table::
   :widths: 1 2
   :header-rows: 0

>    * - **User code fails (**:code:`setup` **or** :code:`run` **commands have non-zero exit code):**

Reviewer: Very nice! However the bolded text is bit hard to read. OK to unbold?

     - If :code:`max_restarts_on_errors` is set, restart up to that many times. If :code:`max_restarts_on_errors` is not set, or we run out of restarts, set the job to :code:`FAILED` or :code:`FAILED_SETUP`.
   * - **Instances are preempted or underlying hardware fails:**
     - Tear down the old temporary cluster and provision a new one in another region, then restart the job.
   * - **Can't find available resources due to cloud quota or capacity restrictions:**
     - Try other regions and other clouds indefinitely until resources are found.
   * - **Cloud config/auth issue or invalid job configuration:**
     - Mark the job as :code:`FAILED_PRECHECKS` and exit. Won't be retried.

To see the logs of user code (:code:`setup` or :code:`run` commands), use :code:`sky jobs logs <job_id>`. If there is a provisioning or recovery issue, you can see the provisioning logs by running :code:`sky jobs logs --controller <job_id>`.

.. tip::

>   Under the hood, SkyPilot uses a "controller" to provision, monitor, and recover the underlying temporary clusters. To learn more about the controller, see :ref:`jobs-controller`.

Reviewer: ```suggestion
  Under the hood, SkyPilot uses a "controller" to provision, monitor, and recover the underlying temporary clusters. See :ref:`jobs-controller`.
```

.. _spot-jobs-end-to-end:

Examples
--------

BERT end-to-end
~~~~~~~~~~~~~~~

> Below we show an `example <https://github.com/skypilot-org/skypilot/blob/master/examples/spot/bert_qa.yaml>`_ for fine-tuning a BERT model on a question-answering task with HuggingFace.

Reviewer: Say we take the BERT YAML from above and make it work xyz

This example:

- has SkyPilot find a V100 instance on any cloud,
- uses spot instances to save cost, and
- uses checkpointing to recover preempted jobs quickly.

> .. code-block:: yaml

Reviewer: Add to beginning of L421 to make it read better "The highlighted lines add in bucket mounting (for saving and reading checkpoints) and instruct the training framework to write checkpoints there." Then make L421 a bit more concise?

periodically -> periodic

(And I know these are remnants from before!)

* PyTorch Lightning DDP, CIFAR-10: `YAML <https://github.com/skypilot-org/skypilot/blob/master/examples/spot/lightning_cifar10.yaml>`__

Running Many Parallel Jobs
--------------------------

For batch jobs such as **data processing** or **hyperparameter sweeps**, you can launch many jobs in parallel. See :ref:`many-jobs`.

> If you want to run more than around 90 jobs at once, you may need to use some advanced configuration. See :ref:`jobs-controller-sizing`.
> 

Reviewer: ```suggestion
To increase the maximum number of running jobs, see :ref:`jobs-controller-sizing`.
```
to be more concise

  recoveries of the job.
  It can be accessed in the task's :code:`run` commands or directly in the program itself (e.g., access
  via :code:`os.environ` and pass to Weights & Biases for tracking purposes in your training script). It is made available to
  the task whenever it is invoked. See more about :ref:`environment variables provided by SkyPilot <sky-env-vars>`.

With the highlighted changes, the managed job can now resume training after preemption! We can enjoy the benefits of
cost savings from spot instances without worrying about preemption or losing progress.

.. code-block:: console

  $ sky jobs launch -n bert-qa bert_qa.yaml

> Real-world examples

Reviewer: Can add @KeplerC 's vector DB workload :-)

  Multiple users can share the same intermediate bucket. Each user's jobs will have their own unique job-specific directories, ensuring that files are kept separate and organized.

.. _jobs-controller:

> How it works: The jobs controller

Reviewer: Does "How it works: Jobs controller" read better?
